rfcs/unsealed-table-assign-indexer.md

```diff
@@ -0,0 +1,61 @@
+# Unsealed table assignment creates an indexer
+
+In Luau, tables have a state, which can, among others, be "unsealed".
+An unsealed table is one that we are still constructing.  Currently
+assigning a table literal to an unsealed table creates new properties,
+but this is a problem in cases where tables are being used as hash
+tables.  We would like to change this so that assigning a table
+literal to an unsealed table creates an indexer rather than
+properties.
+
+## Motivation
+
+In lua-apps, there is testing code which (simplified) looks like:
+
+```lua
+local t = { u = {} }
+t = { u = { p = 37 } }
+t = { u = { q = 5 } }
+local x: number? = t.u["p"]
+local y: number? = t.u["q"]
+```
+
+Currently, this code doesn't typecheck, due to `p` and `q` being unknown properties of `t`.
```

Comment:
Would it be fair to summarize the reason why this code doesn't typecheck as:

- The type checker infers the type of `t.u` to be an empty table
- The assignments of `t` don't change the type because the RHS is compatible with LHS
- As such, access to `t.u` fails.

I don't see why the solution needs to involve string indexers? E.g. if types of `p` and `q` are disjoint here, the problem doesn't seem to be solved correctly by string indexers.

It seemed to me like this is a problem in the general category of "assignments may need to mutate the type of the assigned object". E.g. `local t = nil t = foo` is similar to this example, no?

Reply:
I want to make a separate comment that all these single letter variable names are confusing me as a reader, and I have to frequently dart my eyes back and forth between the code sample to know what you're referring to, I would like it if they had more familiar names.

Reply:
@zeux `t.u` is an unsealed empty table. For unsealed tables, we do allow adding properties, e.g. `t.u.p = 5` is fine, it's just what we should do with assignments of table literals. We could just say "no, not doing that", but this breaks lua-apps https://github.com/Roblox/lua-apps/blob/26c59983268ece9339c5984a2c3d6d6fbabcdab3/src/internal/Social/Modules/RoactChat/Components/ChatConversation/Messages/Selectors/selectMessages.spec.lua#L69-L71

String indexers are relevant because one of the idiomatic uses of table literals is setting a string-indexed table (that's what lua-apps is doing for instance).

Reply:
Naming examples is tricky, The real examples are things like initializing config options.

Reply:
Typing in the comment field is slightly awkward with larger examples so I'll move to the actual comments on the PR

Reply:
Btw @asajeffrey your link doesn't contain an assignment -- the example in that link is a table literal, I don't think there's a table assignment in sight anywhere there.